docs: Add Fluent Bit buffer setup ADR (#574)
kyma-project/telemetry-manager: docs/contributor/arch/002-fluent-bit-without-file-system-buffer.md

# 1. Fluent Bit without File-System Buffer

Date: 2023-11-23

## Status

Proposed

## Context

> The Fluent Bit configuration that we currently generate with the telemetry-manager from a LogPipeline uses a `tail` input plugin to read container logs, splits the log stream into multiple pipelines using a `rewrite_tag` filter, applies additional pipeline specific filters, and sends the log stream to the configured output. This setup decouples the input and outputs so that a blocked output does not affect any other outputs. Consequently, reading logs by the tail plugin is never paused. The `rewrite_tag` filter uses a persistent file-system buffer to prevent log loss. The persistent buffer is limited to 1 GB to ensure that Kubernetes nodes do not run out of disk space.

Reviewer: Simpler. 

```suggestion
The Fluent Bit configuration that the Telemetry Manager currently generates from a LogPipeline uses a `tail` input plugin to read container logs, splits the log stream into multiple pipelines using a `rewrite_tag` filter, applies additional pipeline specific filters, and sends the log stream to the configured output. This setup decouples the input and outputs so that a blocked output does not affect any other outputs. Consequently, reading logs by the tail plugin is never paused. The `rewrite_tag` filter uses a persistent file-system buffer to prevent log loss. The persistent buffer is limited to 1 GB to ensure that Kubernetes nodes do not run out of disk space.
```
Should've noticed this in my first review

2. **Split log stream**: Logs are read by a `tail` input plugin. Before the stream is split into a pipeline-specific stream, global filters like the Kubernetes filter are applied; and before logs are written to an output, each stream can have additional filters. Pipelines can be isolated using the persistent buffer. However, it's impossible to pause the input in the situation of a backend outage. The complexity of this setup is high.
3. **Dedicated log streams**: Each pipeline has its own `tail` input plugin, a list of filters, and an output plugin. This setup isolates the log processing between different pipelines. In the case of a problem, the streams can be paused individually. This setup has medium complexity.
4. **Dedicated Fluent Bit instances**: Each pipeline gets its own Fluent Bit DaemonSet. This setup isolates also the CPU and memory resources per pipeline with the cost of a larger overhead. This setup has medium complexity.

> Option 1 does not fulfill our requirement to apply individual filters per pipeline. Option 4 causes an unacceptable resource overhead for our typical setup of two pipelines (application logs and access logs). Option 2 and 3 allow log filter settings per pipeline. Currently, Telemetry Manager uses option 2.

Reviewer: Seeing it in context, this might be better:

```suggestion
Currently, Telemetry Manager uses option 2. 
Option 1 does not fulfill our requirement to apply individual filters per pipeline. Option 4 causes an unacceptable resource overhead for our typical setup of two pipelines (application logs and access logs). Option 2 and 3 allow log filter settings per pipeline.
```

We consider option 3 to be the best Fluent Bit configuration for our requirements because of its lower complexity and lower risk for log loss. The throughput of option 3 has shown to be better than option 2 without changing Fluent Bit's CPU and memory limits.

## Consequences

> Switching from option 2 to option 3 relies on the container runtime for persistent log storage. This changes the conditions under which logs might be lost. Option 2 with its persistent buffer of 1 GB per pipeline deletes the oldest logs when the limit is reached. For option 3, the conditions under which logs might be lost are specific for the individual pod, because log rotation happens when then log file per pod reaches a certain size. Logs are deleted in the case of pod evictions or multiple restarts. Node deletion might also be a reason for log loss. However, since the persistent buffer of the existing setup uses the host file system, this risk is already existing.

Reviewer: ```suggestion
Switching from option 2 to option 3 relies on the container runtime for persistent log storage. This changes the conditions under which logs might be lost: Option 2, with its persistent buffer of 1 GB per pipeline, deletes the oldest logs when the limit is reached. For option 3, the conditions under which logs might be lost are specific for the individual pod, because log rotation happens when the log file per pod reaches a certain size. Logs are deleted in the case of pod evictions or multiple restarts. Node deletion might also be a reason for log loss. However, because the persistent buffer of the existing setup uses the host file system, this risk is already existing.
```

   and memory resources per pipeline with the cost of a larger overhead. This setup has medium complexity.

Option 1 does not fulfill our requirement to apply individual filters per pipeline. Option 4 will cause for our typical
setup of two pipelines (application logs and access logs) an unacceptable resource overhead. Option 2 and 3 allow log
filter settings per pipeline.

We consider option 3 to be the best Fluent Bit configuration for our requirements because of its lower complexity and
lower risk for log loss. The throughput of option 3 has shown to be better than option 2 without changing Fluent Bit's
CPU and memory limits.

## Consequences

Switching from option 2 to option 3 relies on the container runtime for persistent log storage. This changes the
conditions under which logs might be lost. Option 2 with its persistent buffer of 1 GB per pipeline deletes the
oldest logs when the limit is reached. For option 3, the conditions under which logs might be lost are specific for the
individual pod since log rotation happens when then log file per pod reach a certain size. Logs are deleted in the case

> of pod evictions or multiple restarts.

Reviewer 2: Let's maybe add a word about losing logs when nodes are getting recreated (e.g. if k8s upgrade is performed)

Author: Good catch. Added that case.

# 1. Fluent Bit without File-System Buffer

Date: 2023-11-23

## Status

Proposed

## Context

> The Fluent Bit configuration that we currently generate with the telemetry-manager from a LogPipeline uses a `tail`

Reviewer: let's not split sentences across lines 

```suggestion
The Fluent Bit configuration that we currently generate with the telemetry-manager from a LogPipeline uses a `tail` input plugin to read container logs, splits the log stream into multiple pipelines using a `rewrite_tag` filter, applies additional pipeline specific filters, and sends the log stream to the configured output. This setup decouples the input and outputs so that a blocked output does not affect any other outputs. Consequently, reading logs by the tail plugin is never paused. The `rewrite_tag` filter uses a persistent file-system buffer to prevent log loss. The persistent buffer is limited to 1 GB to ensure that Kubernetes nodes do not run out of disk space.
```

> input plugin to read container logs, splits the log stream into multiple pipelines using a `rewrite_tag` filter, applies

Reviewer: delete the following lines

additional pipeline specific filters, and sends the log stream to the configured output. This setup decouples the input
and outputs so that a blocked output does not affect all other outputs. Consequently, reading logs by the tail plugin is
never paused. The `rewrite_tag` filter uses a persistent file-system buffer to prevent log loss. The persistent buffer
is limited to 1 GB to ensure that Kubernetes nodes do not run out of disk space.

> The persistent buffer can ensure log loss only for a short period of time. Depending on the amount of generated logs,

Reviewer: Please confirm: Does the buffer ENSURE log loss? I thought it prevents log loss.

```suggestion
The persistent buffer can ensure {prevent} log loss only for a short period of time. Depending on the amount of generated logs, this is usually in the range of minutes to a few hours and might be too short to restore an outage of a log backend or detect and solve a faulty configuration.
```

Author: Good catch. It prevents log loss.

this is usually in the range of minutes to a few hours and might be too short to restore an outage of a log backend or
detect and solve a faulty configuration.

> The removal of the in-cluster Loki backend requires a reevaluation of the pipeline isolation requirements in favor of

Reviewer: [we don't say "we"](https://github.com/kyma-project/community/blob/main/docs/guidelines/content-guidelines/02-style-and-terminology.md#pronouns)

```suggestion
After the in-cluster Loki backend has been removed, the pipeline isolation requirements must be reevaluated in favor of pausing the input plugin. The typical setup are clusters with a single LogPipeline or multiple LogPipelines that ingest logs to the same backend; for example, ingesting application logs and Istio access logs into two different indexes of an [SAP Cloud Logging](../../user/integration/sap-cloud-logging) instance. With this typical setup, the benefit of pausing the `tail` input plugin exceeds the pipeline isolation requirements. Kubernetes' [logging architecture](https://kubernetes.io/docs/concepts/cluster-administration/logging/) helps by storing logs persistently on the node file-system and rotating automatically after reaching a certain file size.
```

Kubernetes logging architecture rotates? Or should "rotating" have an object? What is rotated?
Aaah, line 61 says that logs are rotated. Then I'd add a "them":

```suggestion
After the in-cluster Loki backend has been removed, the pipeline isolation requirements must be reevaluated in favor of pausing the input plugin. The typical setup are clusters with a single LogPipeline or multiple LogPipelines that ingest logs to the same backend; for example, ingesting application logs and Istio access logs into two different indexes of an [SAP Cloud Logging](../../user/integration/sap-cloud-logging) instance. With this typical setup, the benefit of pausing the `tail` input plugin exceeds the pipeline isolation requirements. Kubernetes' [logging architecture](https://kubernetes.io/docs/concepts/cluster-administration/logging/) helps by storing logs persistently on the node file-system and rotating them automatically after reaching a certain file size.
```

pausing the input plugin. We consider clusters with a single LogPipeline or multiple LogPipelines that ingest logs to
the same backend as the typical setup. For instance, ingesting application logs and Istio access logs into two different
indexes of a [SAP Cloud Logging](../../user/integration/sap-cloud-logging) instance. With this assumption, the benefit
of pausing the `tail` input plugin exceeds the pipeline isolation requirements. Kubernetes' [logging
architecture](https://kubernetes.io/docs/concepts/cluster-administration/logging/) helps by storing logs persistently on
the node file-system and rotating automatically after reaching a certain file size.

## Decision

Fluent Bit can be configured in different ways to read container logs and ingest them into multiple backends:

> 1. **Single log stream:** Logs are read by a `tail` input plugin, one or more filters are applied, and written to the

Reviewer: ```suggestion
1. **Single log stream:** Logs are read by a `tail` input plugin, one or more filters are applied, and written to the backend by multiple output plugins. The input cannot be controlled based on the condition of individual outputs. Filters cannot be applied per individual pipeline. The complexity of this setup is low.
```

   individual outputs. Filters cannot be applied per individual pipeline. The complexity of this setup is low.
2. **Split log stream**: Logs are read by a `tail` input plugin, global filters like the Kubernetes filter are applied
   before the stream is split into a pipeline specific stream. Each stream can have additional filters before logs are
   written to an output. This setup is currently used by the telemetry-manager. Pipelines can be isolated using the
   persistent buffer. However, pausing the input in the situation of a backend outage is not possible. The complexity of
   this setup is high.
3. **Dedicated log streams**: Each pipeline has its own `tail` input plugin, a list of filters, and output plugin. This
   setup isolates the log processing between different pipelines and allows to pause the streams individually in the
   case of a problem. This setup has medium complexity.
4. **Dedicated Fluent Bit instances**: Each pipeline gets its own Fluent Bit DaemonSet. This setup isolates also the CPU
   and memory resources per pipeline with the cost of a larger overhead. This setup has medium complexity.

> Option 1 does not fulfill our requirement to apply individual filters per pipeline. Option 4 will cause for our typical

Reviewer: ```suggestion
Option 1 does not fulfill our requirement to apply individual filters per pipeline. Option 4 causes an unacceptable resource overhead for our typical setup of two pipelines (application logs and access logs). Option 2 and 3 allow log filter settings per pipeline. Currently, Telemetry Manager uses option 2.
```

setup of two pipelines (application logs and access logs) an unacceptable resource overhead. Option 2 and 3 allow log
filter settings per pipeline.

> We consider option 3 to be the best Fluent Bit configuration for our requirements because of its lower complexity and

Reviewer: ```suggestion
We consider option 3 to be the best Fluent Bit configuration for our requirements because of its lower complexity and lower risk for log loss. The throughput of option 3 has shown to be better than option 2 without changing Fluent Bit's CPU and memory limits.
```

lower risk for log loss. The throughput of option 3 has shown to be better than option 2 without changing Fluent Bit's
CPU and memory limits.

## Consequences

> Switching from option 2 to option 3 relies on the container runtime for persistent log storage. This changes the

Reviewer: ```suggestion
Switching from option 2 to option 3 relies on the container runtime for persistent log storage. This changes the conditions under which logs might be lost. Option 2 with its persistent buffer of 1 GB per pipeline deletes the oldest logs when the limit is reached. For option 3, the conditions under which logs might be lost are specific for the individual pod, because log rotation happens when then log file per pod reaches a certain size. Logs are deleted in the case of pod evictions or multiple restarts. Node deletion might also be a reason for log loss. However, since the persistent buffer of the existing setup uses the host file system, this risk is already existing.
```
